### PR TITLE
Fixes a panic when migrating hibernated shoots

### LINKED ISSUE #3759: panic: runtime error: invalid memory address or nil pointer dereference
**How to categorize this issue?**

/area control-plane-migration
/kind bug
/priority 3

**What happened**:

gardenlet is `CrashLoopBackOff` because of a panic in the controlplane migration flow.

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x60 pc=0x1d18769]

goroutine 4641 [running]:
github.com/gardener/gardener/pkg/operation/botanist.(*Botanist).setAPIServerServiceClusterIP(0xc003ac3ec0, 0xc00ee1c230, 0xc)
	/go/src/github.com/gardener/gardener/pkg/operation/botanist/controlplane.go:987 +0x289
github.com/gardener/gardener/pkg/operation/botanist/controlplane.(*kubeAPIService).Deploy(0xc00f02a0b0, 0x25bb340, 0xc000056018, 0xc00f02a0b0, 0x0)
	/go/src/github.com/gardener/gardener/pkg/operation/botanist/controlplane/kube_apiserver_service.go:138 +0x255
github.com/gardener/gardener/pkg/operation/botanist.(*Botanist).DeployKubeAPIService(0xc003ac3ec0, 0x25bb340, 0xc000056018, 0x1, 0x0, 0x0)
	/go/src/github.com/gardener/gardener/pkg/operation/botanist/controlplane.go:953 +0x5d
github.com/gardener/gardener/pkg/operation/botanist.(*Botanist).WakeUpKubeAPIServer(0xc003ac3ec0, 0x25bb340, 0xc000056018, 0xa46c8bc, 0xc002ee7da0)
	/go/src/github.com/gardener/gardener/pkg/operation/botanist/controlplane.go:213 +0x85
github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode.func1(0xc0077749a0, 0x2326781, 0x38, 0x25bb340, 0xc000056018)
	/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:219 +0x1d5
created by github.com/gardener/gardener/pkg/utils/flow.(*execution).runNode
	/go/src/github.com/gardener/gardener/pkg/utils/flow/flow.go:214 +0x14f
```

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:

- Gardener version:

```
version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.1", GitCommit:"fb648412afde6ff8a1e8d1f92b4378f3c88d8d35", GitTreeState:"clean", BuildDate:"2021-03-22T14:49:03+00:00", GoVersion:"go1.15.9", Compiler:"gc", Platform:"linux/amd64"}
```

- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

### PR DESCRIPTION
**Pull request by plkokanov**
**How to categorize this PR?**

/area control-plane-migration
/kind bug

**What this PR does / why we need it**:
This PR fixes а nil pointer exception when migrating a hibernated shoot and SNI is enabled.

**Which issue(s) this PR fixes**:
Fixes #3759 

**Special notes for your reviewer**:
When using SNI the `botanist.SeedNamespaceObject` has to be set as [this](https://github.com/gardener/gardener/blob/4f729681c7635df9e2ceacd55e1d581b5dd73a7b/pkg/operation/botanist/controlplane.go#L935) code is called when `botanist.WakeUpKubeAPIServer` is executed during migration of a hibernated shoot.

**Release note**:

```other operator
Fixes a nil pointer dereference panic that happens when migrating a hibernated shoot and SNI is enabled.
```

### CODE REVIEW

File: pkg/gardenlet/controller/shoot/shoot_control_migrate.go [hunk 1]
```diff
@@ -126,7 +126,8 @@ func (c *Controller) runPrepareShootControlPlaneMigration(o *operation.Operation
 			return nil
 		}),
 		utilerrors.ToExecute("Retrieve the Shoot namespace in the Seed cluster", func() error {
-			if err := botanist.K8sSeedClient.Client().Get(ctx, client.ObjectKey{Name: o.Shoot.SeedNamespace}, namespace); err != nil {
+			botanist.SeedNamespaceObject = &corev1.Namespace{}
+			if err := botanist.K8sSeedClient.Client().Get(ctx, client.ObjectKey{Name: o.Shoot.SeedNamespace}, botanist.SeedNamespaceObject); err != nil {
```
Comment (vpnachev) [inline pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: after this if block, you need something like 
```golang
namespace = botanist.SeedNamespaceObject
```

**Review by vpnachev — CHANGES_REQUESTED** [review]

File: pkg/gardenlet/controller/shoot/shoot_control_migrate.go (hunk 1 above)
Comment (ialidzhikov) [inline pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: Isn't the better approach to add `.DoIf(nonTerminatingNamespace)` to the the task that causes the panic (`botanist.WakeUpKubeAPIServer` in this case)?

Wouldn't the following fix the issue?

```diff
diff --git a/pkg/gardenlet/controller/shoot/shoot_control_migrate.go b/pkg/gardenlet/controller/shoot/shoot_control_migrate.go
index 06d06dbc9..b0ac9e6fb 100644
--- a/pkg/gardenlet/controller/shoot/shoot_control_migrate.go
+++ b/pkg/gardenlet/controller/shoot/shoot_control_migrate.go
@@ -183,7 +183,7 @@ func (c *Controller) runPrepareShootControlPlaneMigration(o *operation.Operation
                })
                wakeUpKubeAPIServer = g.Add(flow.Task{
                        Name:         "Scaling Kubernetes API Server up and waiting until ready",
-                       Fn:           flow.TaskFn(botanist.WakeUpKubeAPIServer).DoIf(wakeupRequired),
+                       Fn:           flow.TaskFn(botanist.WakeUpKubeAPIServer).DoIf(wakeupRequired && nonTerminatingNamespace),
                        Dependencies: flow.NewTaskIDs(deployETCD, scaleETCDToOne),
                })
                ensureResourceManagerScaledUp = g.Add(flow.Task{
```

Comment (plkokanov) [reply pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: `wakeupRequired` is only true if `nonTerminatingNamespace` is already true as well. 
```
cleanupShootResources   = nonTerminatingNamespace && kubeAPIServerDeploymentFound
wakeupRequired          = (o.Shoot.Info.Status.IsHibernated || (!o.Shoot.Info.Status.IsHibernated && o.Shoot.HibernationEnabled)) && cleanupShootResources
```
The nil pointer exception happens always when SNI is enabled and the shoot is hibernated (not only when the namespace is being already deleted)

Comment (plkokanov) [reply pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: I'll just get rid of the `namespace` var and use
```
nonTerminatingNamespace = botanist.SeedNamespaceObject.Status.Phase != corev1.NamespaceTerminating
```

seems like the better approach

**Comment by plkokanov** [comment]
Applied the review comments @vpnachev @ialidzhikov could you take another look.

File: pkg/gardenlet/controller/shoot/shoot_control_migrate.go [hunk 2]
```diff
@@ -126,7 +125,8 @@ func (c *Controller) runPrepareShootControlPlaneMigration(o *operation.Operation
...
```
Comment (rfranzke) [inline pkg/gardenlet/controller/shoot/shoot_control_migrate.go]:
```suggestion
			if err := botanist.K8sSeedClient.APIReader().Get(ctx, client.ObjectKey{Name: o.Shoot.SeedNamespace}, botanist.SeedNamespaceObject); err != nil {
```

According to https://github.com/gardener/gardener/blob/b5b0fe3f91d71991d6649518e7d2a12246f18474/pkg/gardenlet/controller/shoot/shoot_control_delete.go#L88-L98 :)

Comment (vpnachev) [reply pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: If this applied, I guess it should be done also for https://github.com/gardener/gardener/blob/3a394eeb705f1814e27053f4ed618d5dbe95c62d/pkg/gardenlet/controller/shoot/shoot_control_migrate.go#L116

**Review by vpnachev — DISMISSED** [review]
/lgtm

File: pkg/gardenlet/controller/shoot/shoot_control_migrate.go (hunk 2 above)
Comment (rfranzke) [reply pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: Yep, nice catch @vpnachev, thanks!

Comment (plkokanov) [reply pkg/gardenlet/controller/shoot/shoot_control_migrate.go]: Also noticed that `nil` was returned at the end of the function even if reading the namespace returned an error different than `NotFound`.

**Comment by plkokanov** [comment]
Applied the new review change requests :).

**Review by vpnachev — APPROVED** [review]
/lgtm

**Pull request merged**